### PR TITLE
PaintOnTemplateTool: Add filled area mode

### PR DESCRIPTION
**Pull request by lpechacek**
The `ScribbleOptions` part may feel like YAGNI but I have another idea for an enhancement. So at this point we can go with a boolean parameter to the method or leave the more descriptive and extensible `QFlags` in place. Please review, Kai.

### CODE REVIEW

File: src/templates/template.h [hunk 1]
```diff
@@ -141,6 +141,7 @@ Q_OBJECT
 	 */
 	enum ScribbleOption
 	{
+		NoScribbleOptions = 0,
```
Comment (lpechacek) [inline src/templates/template.h:144]: Yes, I had this line in the code but then I removed it. The "no options" is equivalent to newly initialized `QFlags` or just plain `{}`. That is why I removed the "no bits set" before submission.

Your proposal made me look at `QFlags` usage in Qt (`fgrep -B15 -r Q_DECLARE_FLAGS /usr/include/qt5/`). I can see that zero appears in the enums in about half of the cases. That said, I don't mind re-introduction of the value. This comment is a side note explaining why it was "missing".

File: src/templates/template.h [hunk 2]
```diff
@@ -744,9 +745,10 @@ Q_OBJECT
 	Matrix template_to_map_other;
 };
 
-Q_DECLARE_OPERATORS_FOR_FLAGS(Template::ScribbleOptions)
-
 
 }  // namespace OpenOrienteering
 
+Q_DECLARE_OPERATORS_FOR_FLAGS(OpenOrienteering::Template::ScribbleOptions)
```
Comment (lpechacek) [inline src/templates/template.h:751]: Ack this one.

**Comment by lpechacek** [comment]
Thanks for the review, Kai.

Ad the pen widths, I believe that the filled areas mode provides a more capable alternative. As M. Lerjen [mentioned in the associated issue](https://github.com/OpenOrienteering/mapper/issues/1242#issuecomment-489554310), the fill areas option allows creating colored patches with precisely defined edges. Thicker pens do not provide this benefit.

My idea for enhancement was "overprint" mode. Recently I needed to overlay a color area over a few hand-sketched symbols. Combining these elements turned out to be quite tedious. I also think that the "overprint" scribbling mode could help people migrate from Mylar to a tablet.

Just my two cents, I'll be happy to read alternative opinions. :)

P.S. Plus a richer color palette would be required for migration from Mylar. Even I used about 9 pencils. Common number of pencils is 10-15 ([according to The Czech Mapper's Tutorial](http://tvorbamap.shocart.cz/vybaveni/soucasti.htm#tuzky)).

File: src/templates/template.h (hunk 1 above)
Comment (dg0yt) [reply src/templates/template.h:144]: Yes, this is is just a name for `ScribbleOptions{}`. However, a single-valued enum feels strange to mee, at least outside of C++ meta programming.

**Comment by mlerjen** [comment]
image
Before. Tendency to incomplete area work.
image
With Fill areas in scribble mode.

The new filll area mode is a nice  and appreciated enhancement leading to much better field notes. Areas now can easily be represented by entirely filled areas instead of  their outlines  and a scribbled (time consuming) indication of the filling. A fine pen with and a efficient area filling are no longer contradicting.

image
So why not use the fill areas mode here?

In the logic of the workflow though, you would first draw point and line features making use of the template(s) before concluding a section with the specification of the area.  As the filling areas mode covers/erases all scribble it thus turns unhandy. In the example above the tool could only be applied to to the various sections of the central brown area each confined by either area outline or contours. (actually that was done for the upper brown area).

I would though consider to make the fill area mode "filling" instead of "covering" (replacing the only the transparent parts of a pixel with the filling color?? Except for white still voiding everything).

image
Point and short line features drawn directly. Contours as vectors sketched with a helper symbol and drawn with the freehand tool. Areas with out-lines on the scribble layer.

As a work-around to preserve the strength of the tool as it is now ("covering"). you can use the "draw freehand tool" with a helper line symbol to draw line features. These will not be covered by the concluding "fill area" scribble. This would btw. also be a way to approximate the full colorful potential of the Czech Map Painting tradition ;-)

**Comment by dg0yt** [comment]
> I would though consider to make the fill area mode "filling" instead of "covering" (replacing the only the transparent parts of a pixel with the filling color?? Except for white still voiding everything).

Point taken. However, this would mean that you cannot re-color any area with the fill tool, unless erasing first. Or unless we find a way to handle at least two "layers".

**Comment by lpechacek** [comment]
> > I would though consider to make the fill area mode "filling" instead of "covering" (replacing the only the transparent parts of a pixel with the filling color?? Except for white still voiding everything).
>
> Point taken. However, this would mean that you cannot re-color any area with the fill tool, unless erasing first. Or unless we find a way to handle at least two "layers".

The implementation is quite straightforward (https://github.com/lpechacek/mapper/commit/0aedd4be91f3b78c47da4044fb84deb9318de010) and I think the restriction @dg0yt mentions is far less hassle than trying to fit new drawing between existing lines.

I'd try producing a test build but my Pipelines instance is not yet configured. For a while I thought about creating a new branch in the official OO repo but then I realized that a (draft) pull request will also trigger Pipelines build with the possibility to download the signature-less test APK.

**Comment by lpechacek** [comment]
A (self-)signed test APK is at http://liborpechacek.wz.cz/oomapper/OpenOrienteering-Mapper-refs_pull_1660_merge_v20200619.1-Android-armeabi-v7a.apk, however, I'm far from recommending installation of the package, given the previous experience with test builds.

The new tool behavior feels slightly counterintuitive to me. Making changes to existing drawing just does not "stick". The pen leaves a trace only in transparent areas.

**Comment by dg0yt** [comment]
> The new tool behavior feels slightly counterintuitive to me. Making changes to existing drawing just does not "stick". The pen leaves a trace only in transparent areas.

I think this is what I meant by

> you cannot re-color any area with the fill tool, unless erasing first.

but extended to the pen!

Apart from the layers approach I mentioned, options are
- a pattern (like during drawing)
- multiplying blend mode for all drawing actions (like in overprinting simulation).

**Comment by mlerjen** [comment]
Screenshot_20200619-180953_OpenOrienteering Mapper refspull1660merge v202006191

Great fun from my point of view. Last things to fine-tune maybe: 
- while the line pen has a certain opacity, the area pens are opaque and do blend with the drawn lines. (e.g. red and green result into a rather blackish line). Features scribbled with the line features thus tend to be "dissolved" by the areas.
- the button for the "fill area"... hm ... ?? :-)

Thanks again!

**Comment by dg0yt** [comment]
> * while the line pen has a certain opacity, the area pens are opaque and do blend with the drawn lines.

This is the "multiplying blend mode" I mentioned.

**Comment by lpechacek** [comment]
I've kept silent because I have no smart idea of how to solve the insufficiencies at the moment.

Let me at least draft the options:
1) The overprint/"multiplying blend" mode was my first idea. It is supposed to match pencil drawing but results in hue changes of the existing drawing. We haven't tried it so far, but I'm more and more persuaded that we should do give a shot to this one.

2) "Destination over" mode - current implementation where existing drawing projects through the new additions. There is a change in hue at the edges, where the drawing (alpha-) blends into the background. For lines, it means that they are less visible but the centerline still keeps its original hue. We could solve it by switching antialiasing off, but I'm not sure what would be the end-user response. The drawing would give an uglier appearance. Alternatively, thicker line would solve this as well.

Based on the increasing demand, I'll also look into the fill tool. :) My only concern is the increasing number of buttons in the scribble toolbar.

**Comment by ollesmaps** [comment]
> My only concern is the increasing number of buttons in the scribble toolbar.  

What about a collapsible toolbar as suggested in issue #947. This isue was raised 3 years ago, when I was using small phone screen. Now I have a bigger one and it is not so pressing, however some might find it handy.

This also relates to color choice in scribble toolbar issue #945.

**Comment by dg0yt** [comment]
> What about a collapsible toolbar as suggested in issue #947.

There is much non-standard code around the scribbling tools. I collected all related issues in a ["project"](https://github.com/OpenOrienteering/mapper/projects/7). (This PR is not added to the project, but the original issue is.) Now we still need someone to work on that.

**Comment by lpechacek** [comment]
> There is much non-standard code around the scribbling tools. I collected all related issues in a "project". (This PR is not added to the project, but the original issue is.) Now we still need someone to work on that.

Ack. I'm aware of the project and read through it today again. I've got some idea about how to implement the requests, and I'd consider doing it for the pure joy of enhancing Mapper functionality. I'm, however, unsure what you mean with non-standard code and if you expect future contributions to move the code (style) in a certain direction.

**Comment by dg0yt** [comment]
This is the result of painting with [`Qt::Dense4Pattern`](https://doc.qt.io/qt-5/qt.html#BrushStyle-enum), leaving some pixels untouched: 
dense-pattern

- It might make sense to turn off the outline (Qt: "pen") when drawing with a pattern, because repetitions will leave covering lines.
- On the same image, the pattern is always applied to the same pixels. Repeated painting with the same pattern will always change the same pixels (to the current color). No darkening, no mixing - only pixels created by pens or solid fills will shine through.
- 1 pixel wide pen strokes at 45° may get covered completely. (Cf. green line below red area.)

**Pull request merged**